MSC2409: Proposal to send EDUs to appservices (#2409)
matrix-org/matrix-spec-proposals: proposals/2409-appservice-edus.md

Reviewer: As to-device messages are often talked about in the context of EDUs, could you explicitly say that this MSC **does not** cover to-device events and that it is left to a future MSC?

# MSC2409: Proposal to send EDUs to appservices

*Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
and deprecates that one.*

> The appservice /transaction API currently supports pushing PDU events (regular message and state events)

Reviewer 2: link to the API please

Author: Done!

however it doesn't provison for EDU events (typing, presence and more). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal

> specifiys how these can be pushed to an appservice.

Reviewer 2: ```suggestion
specifies how these can be pushed to an appservice.
```

         "type": "m.room.member",
         "event_id": "$143273582443PhrSn:domain.com",
         "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
         "sender": "@example:domain.com",
         "origin_server_ts": 1432735824653,
         "unsigned": {
            "age": 1234
         },
         "state_key": "@alice:domain.com"
      }
   ]
}
```

This proposal would extend the `PUT /_matrix/app/v1/transactions/` endpoint to include a new key
`ephemeral` to match the CS APIs `/sync`.

```json
{
   "ephemeral": [
      {
         "type": "m.typing",
         "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
         "content": {
            "user_ids": [
               "@alice:example.com"
            ]
         }
      },
      {
         "type": "m.receipt",
         "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
         "content": {
            "$1435641916114394fHBLK:matrix.org": {
               "m.read": {
                  "@rikj:jki.re": {
                     "ts": 1436451550453
                  }
               }
            }
         }
      }
   ],
   "events": [
       // ...
   ]
}
```

The reason for a new key rather than bundling the events into `events` is that

> old implementations may mistake them for PDUs and could cause undefined behaviour.

Reviewer 2: ```suggestion
existing appservices may mistake them for PDUs and could cause undefined behaviour.
```

While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. An appservice
registers interests in rooms and (usually) it's own users, however EDU events are not
tied to a single room in all situtations and as such there needs to be a specified way of
forwarding these events.

An EDU should be sent to an appservice if the `room_id` is shared by any of the registered appservices
users, if possible. For EDUs where that isn't the case, that is `m.presence`, the EDU should be sent
if the sender is present in a room that is shared by any of the registered appservices users.

## Potential issues

Determining which EDUs to transmit to the appservice could lead to quite some overhead on the
homeservers side. Additionally, more network traffic is produced, potentially straining the local

> network and the appservice more.

Reviewer 2: due to the performance issues involved, the registration should have an opt-in flag so the homeserver can skip processing the EDUs if the appservice doesn't care. A later MSC can provide more granular filtering on both PDUs and EDUs.

Author: added `receive_edus` to the registration file!

> # MSC2409: Proposal to send EDUs to appservices

Reviewer 3: Having thought about this a bit more, one of the difficult parts of implementation (for me) has been trying to shoe-horn EDU traffic into transactions. Transactions are logical blocks of events that should be retried if the service is down, but EDUs are categorically not (in most cases).

If you miss a presence, read receipt or typing then you typically as a service don't want that to be retained. So in some ways this feels like the wrong shape of API, since we're doing a `PUT /transaction/12345` except we don't expect 12345 to contain the same contents anymore.

So I am wondering if there should be a seperate endpoint to push this traffic down, which just pushes whenever the homeserver has data (AND the appservice is considered "up"). 

(to_device is a bit of a special case in that homeservers DO want to send those reliably, so that may need to come down a third endpoint, or shoved into a transaction where a transaction ID roughly maps to the stream ID of an inbox)

Author: how is it so much different from a `/sync`?

Reviewer 4: `/sync` doesn't reliably deliver everything like appservice transactions do

Author: Then...soru doens't quite understand where the problem is. You try to put a transaction, it fails, you re-try later. Why does that re-trying later have to contain the same edus? what is stopping it from just...not containing them?

Reviewer 5: This problem isn't really any different to the desired unreliability of those EDUs over federation, and federation uses transactional endpoints too, right?

Reviewer 3: >You try to put a transaction, it fails, you re-try later. Why does that re-trying later have to contain the same edus?

I could be reading the spec wrong, but the txnId from the homeserver basically ensures that the content in the transaction is going to be the same for the same txnId. Ergo if you send txn 123, and it fails, and then you resend it with different EDU data, then you've violated that promise. 

>And even if it has to contain the same edus, why is that bad

I think because typically it's unhelpful for appservices to receive stale presence/typing data, and potentially read receipts too since they get bundled as one event.

On the other hand, we can choose not to care about any of this for the EDU field, but it does feel a bit wrong.

Reviewer 4: > I could be reading the spec wrong, but the txnId from the homeserver basically ensures that the content in the transaction is going to be the same for the same txnId. Ergo if you send txn 123, and it fails, and then you resend it with different EDU data, then you've violated that promise.

This is a Matrix spec change, which means we can change the spec. That promise doesn't necessarily have to apply to EDUs.

Reviewer 2: it would go against the terminology of Matrix to change the behaviour for a single transaction endpoint. EDUs should be reliably known by the server given they are repeated in federated transactions, so there's not really an excuse to drop them on the floor.

Reviewer 3: The synapse implementation uses /transactions to do this, so closing.

*Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
and deprecates that one.*

The [appservice /transactions API](https://matrix.org/docs/spec/application_service/r0.1.2#put-matrix-app-v1-transactions-txnid)
currently supports pushing PDU events (regular message and state events)
however it doesn't provison for EDU events (typing, presence and more). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal
specifies how these can be pushed to an appservice.

## Proposal

### Changes to the registration file

In order that appservices don't get flooded with EDUs, appservices have to opt-in to receive them by

> setting `receive_edus` to true. A registration file could look like following:

Reviewer 3: Can we change this to `push_ephemeral`, my reasons are twofold:
- This is an instruction for the homeserver to push.
- The bridge won't receive raw EDUs in a technical sense, but rather the resulting C-S API edition of the event (e.g. some events are manufactured in the homeserver like read_receipts).

as_token: "30c05ae90a248a4188e620216fa72e349803310ec83e2a77b34fe90be6081f46"
hs_token: "312df522183efd404ec1cd22d2ffa4bbc76a8c1ccf541dd692eef281356bb74e"
sender_localpart: "_irc_bot"
# We want to receive EDUs
receive_edus: true
namespaces:
  users:
    - exclusive: true
      regex: "@_irc_bridge_.*"
  aliases:
    - exclusive: false
      regex: "#_irc_bridge_.*"
  rooms: []
```

### Changes to the /transactions/ API

The `PUT /_matrix/app/v1/transactions/{txnId}` API currently supports sending PDUs
via the `events` array.

```json
{
  "events": [
    {
      "content": {
        "membership": "join",
        "avatar_url": "mxc://domain.com/SEsfnsuifSDFSSEF#auto",
        "displayname": "Alice Margatroid"
      },
      "type": "m.room.member",
      "event_id": "$143273582443PhrSn:domain.com",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "sender": "@example:domain.com",
      "origin_server_ts": 1432735824653,
      "unsigned": {
        "age": 1234
      },
      "state_key": "@alice:domain.com"
    }
  ]
}
```

This proposal would extend the `PUT /_matrix/app/v1/transactions/` endpoint to include a new key
`ephemeral` to match the CS APIs `/sync`.

```json
{

>   "ephemeral": [

Reviewer 3: Presumably this will need a further MSC, but it would be good to be able to send device_list changes to the appservice which aren't EDUs at all?

Reviewer 2: that will probably need a new MSC indeed. Appservices don't *really* have real devices they can rely on in some server implementations.

Reviewer 2: Other MSCs seem to give the appservice a device reliably, so I think this MSC can now adopt EDUs properly.

existing appservices may mistake them for PDUs and could cause undefined behaviour.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. An appservice
registers interests in rooms and (usually) it's own users, however EDU events are not
tied to a single room in all situations and as such there needs to be a specified way of
forwarding these events.

An EDU should be sent to an appservice if the `room_id` is shared by any of the registered appservices
users, if possible. For EDUs where that isn't the case, that is `m.presence`, the EDU should be sent
if the sender is present in a room that is shared by any of the registered appservices users.

## Potential issues

Determining which EDUs to transmit to the appservice could lead to quite some overhead on the
homeservers side. Additionally, more network traffic is produced, potentially straining the local
network and the appservice more. As such, appservices have to opt-in to receive EDUs.

> ## Security considerations

Reviewer 3: @Sorunome it would be good to have an unstable prefix section so I can implement this. We would need one on the transaction body `ephemeral` key, so something like `de.sorunome.msc2409.ephemeral`. We would also need one for the registration file key `receive_edus` (`de.sorunome.msc2409.receive_edus`).

Reviewer 3: As per the above, keen we use push_ephemeral

Reviewer 3: So if we could have a `de.sorunome.msc2409.push_ephemeral` key, and a `de.sorunome.msc2409.ephemeral` for the transaction body that would be good :+1:

# MSC2409: Proposal to send EDUs to appservices

*Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
and deprecates that one.*

The [appservice /transactions API](https://matrix.org/docs/spec/application_service/r0.1.2#put-matrix-app-v1-transactions-txnid)
currently supports pushing PDU events (regular message and state events)
however it doesn't provison for EDU events (typing, presence and more). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal

> specifies how these can be pushed to an appservice.

Reviewer 2: another justification is that complex bots might find this information valuable, but instead have to weigh a risk of either crushing the server with `/sync` or not having the relevant features.

## Proposal

### Changes to the registration file

> In order that appservices don't get flooded with EDUs, appservices have to opt-in to receive them by
> setting `push_ephemeral` to true. A registration file could look like following:

Reviewer 2: what if it is set to false?

```suggestion
To prevent accidentally flooding existing appservices with EDUs, appservices must set `push_ephemeral` to
`true` in their registration. When `false` (the default), the appservice will not receive EDUs. A registration file 
could look like following:
```

Reviewer 2: hello november travis: this is pretty obvious.

          "@alice:example.com"
        ]
      }
    },
    {
      "type": "m.receipt",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that
existing appservices may mistake them for PDUs and could cause undefined behaviour.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. An appservice
registers interests in rooms and (usually) it's own users, however EDU events are not
tied to a single room in all situations and as such there needs to be a specified way of
forwarding these events.

An EDU should be sent to an appservice if the `room_id` is shared by any of the registered appservices
users, if possible. For EDUs where that isn't the case, that is `m.presence`, the EDU should be sent

> if the sender is present in a room that is shared by any of the registered appservices users.

Reviewer 2: This is slightly different from today's specification. Today's specification uses a concept of interest to determine if an event needs to be sent to the application service. Registration of a user is not required in order to be interested in an event - the user just has to be in the namespace. As such, this paragraph could be reworded as:

> An appservice is interested in an EDU if the appservice would be interested in the associated room or sender as though it were a regular room event. For example, the presence of a user (not associated with a room) would be sent to the appservice if the appservice had a namespaced user participating in a room with that user or the room ID/aliases match the appservice's namespace(s). Similarly, a typing event (which is associated with a room) would be sent if the appservice was interested in the room or sender.

`ephemeral` to behave similar to the various sections of the CS API `/sync` endpoint. The `ephemeral` key
MAY be omitted entirely if there are no ephemeral events to send.

```json
{
  "ephemeral": [
    {
      "type": "m.typing",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "user_ids": [
          "@alice:example.com"
        ]
      }
    },
    {
      "type": "m.receipt",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that
existing appservices may mistake them for PDUs and might behave erratically.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

The array is effectively a combination of the `presence` and `ephemeral` sections of the
client-server `/sync` API. User-defined ephemeral events don't exist yet, which means there are

> only three event types that can currently occur: `m.presence`, `m.typing` and `m.receipt`.

Reviewer 6: I've thought a bit about it and don't see any issues, but wanted to double check that you considered both private read receipts and threaded read receipts when re-evaluating this proposal?

EDUs are formatted the same way as they are in C-S sync, with the addition of the `room_id` field
for room-scoped EDUs (`m.typing` and `m.receipt`). `room_id` is not present in the C-S API because
sync nests EDUs inside a room object, but appservices get a flat list of events in all rooms.

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. Appservices claim
namespaces of users which registers "interest" in the rooms where those users reside, as
well as claiming namespaces of rooms for explicit interest. However, not all EDUs are
associated with a single room (presence, etc).

If the EDU is capable of being associated to a particular room (i.e. `m.typing` and `m.receipt`), it should be sent to the

> appservice under the same rules as regular events (interest in the room means sending it).

Reviewer 4: Hmm, I guess it should only send if the appservice is interested in the user

> # MSC2409: Proposal to send EDUs to appservices

Reviewer 7: Might be good to clarify that this no longer covers all EDUs

```suggestion
# MSC2409: Proposal to send typing, presence and receipts to appservices
```

`ephemeral` to behave similar to the various sections of the CS API `/sync` endpoint. The `ephemeral` key
MAY be omitted entirely if there are no ephemeral events to send.

```json
{
  "ephemeral": [
    {
      "type": "m.typing",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "user_ids": [
          "@alice:example.com"
        ]
      }
    },
    {
      "type": "m.receipt",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that
existing appservices may mistake them for PDUs and might behave erratically.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

The array is effectively a combination of the `presence` and `ephemeral` sections of the
client-server `/sync` API. User-defined ephemeral events don't exist yet, which means there are

> only three event types that can currently occur: `m.presence`, `m.typing` and `m.receipt`.

Reviewer 7: ```suggestion
only three event types that can currently occur: [`m.presence`](https://spec.matrix.org/v1.11/client-server-api/#mpresence), [`m.typing`](https://spec.matrix.org/v1.11/client-server-api/#mtyping), and [`m.receipt`](https://spec.matrix.org/v1.11/client-server-api/#mreceipt).
```

# MSC2409: Proposal to send typing, presence and receipts to appservices

> *Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)

Reviewer: ```suggestion
*Note: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
```

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. Appservices claim
namespaces of users which registers "interest" in the rooms where those users reside, as
well as claiming namespaces of rooms for explicit interest. However, not all EDUs are
associated with a single room (presence, etc).

If the EDU is capable of being associated to a particular room (i.e. `m.typing` and `m.receipt`),
it should be sent to the appservice under the same rules as regular events (interest in the room
means sending it). For EDUs which are not associated with a particular room, the appservice
receives the EDU if it contextually *would* apply. For example, a presence update for a user an
appservice shares a room with (or is under the appservice's namespace) would be sent to the
appservice.

For `m.receipt`, private read receipts (`m.read.private`) should only be sent for users within the
appservice's namespaces. Normal read receipts and threaded read receipts are always sent.

## Potential issues

Determining which EDUs to transmit to the appservice could lead to quite some overhead on the

> homeservers side. Additionally, more network traffic is produced, potentially straining the local

Reviewer: ```suggestion
homeserver side. Additionally, more network traffic is produced, potentially straining the local
```

network and the appservice more. As such, appservices have to opt-in to receive EDUs.

## Security considerations

The homeserver needs to accurately determine which EDUs to send to the appservice, as to not leak

> any metadata about users. Particularly `m.presence` could be tricky, as no `room_id` is present in

Reviewer: ```suggestion
any unnecessary metadata about users. Particularly `m.presence` could be tricky, as no `room_id` is present in
```

and deprecates that one.*

The [appservice /transactions API](https://spec.matrix.org/v1.11/application-service-api/#put_matrixappv1transactionstxnid)
currently supports pushing PDU events (regular message and state events)
however it doesn't provision for EDU events (typing, presence and receipts). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal
specifies how these can be pushed to an appservice.

## Proposal

### Changes to the registration file

In order that appservices don't get flooded with EDUs, appservices have to opt-in to receive them by
setting `receive_ephemeral` to true. A registration file could look like following:

```yaml
id: "IRC Bridge"
url: "http://127.0.0.1:1234"
as_token: "30c05ae90a248a4188e620216fa72e349803310ec83e2a77b34fe90be6081f46"
hs_token: "312df522183efd404ec1cd22d2ffa4bbc76a8c1ccf541dd692eef281356bb74e"
sender_localpart: "_irc_bot"
# We want to receive EDUs

> receive_ephemeral: true

Reviewer 8: [non-blocking comment:] Might be nice to allow some way of selecting *which* EDUs to receive, but that could be something added in a future MSC.

Reviewer 4: There were at least two threads and one non-thread discussion about filtering, but somehow it never ended up being mentioned in the MSC. I think the general consensus is that it's better to leave to a future MSC.

# MSC2409: Proposal to send EDUs to appservices

*Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
and deprecates that one.*

The [appservice /transactions API](https://matrix.org/docs/spec/application_service/r0.1.2#put-matrix-app-v1-transactions-txnid)
currently supports pushing PDU events (regular message and state events)
however it doesn't provison for EDU events (typing, presence and more). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal
specifies how these can be pushed to an appservice.

## Proposal

### Changes to the registration file

In order that appservices don't get flooded with EDUs, appservices have to opt-in to receive them by

> setting `push_ephemeral` to true. A registration file could look like following:

Reviewer 9: Is this granular enough? What if an appservice wants to receive typing notifications but not presence updates? It could be a significant resource savings to avoid sending types of events that are just going to be thrown away. Something as simple of making this a list of types could save a lot of resources. Maybe for future extensibility it could be a map of type to empty object, a default can be specified with the type of `*` (or a separate key). I imaging that most appservices would have a static list of events that they support so listing the supported event types makes sense. I suspect that getting all events is a rarer case.

On the other hand the config file is kinda the wrong place for this, maybe we can let the appservice negotiate with the homeserver somehow. However in that case why do we have a flag in the config file as well? Why not just let the appservice tell the server which events it is interested in (and it can remind/update the server every time it boots up).

I see that this was discussed before (https://github.com/matrix-org/matrix-doc/pull/2409#issuecomment-602075652) but I think we should reconsider this. I would rather not have a flag in the config file as it isn't something that an admin would (normally) care about. If we are going to add filtering in the future why don't we add that as the opt-in mechanism?

Reviewer 2: Filtering on the server side wouldn't be much more efficient given the proximity appservices have to the inner workings of the homeserver. Much like room events, the appservice can ignore whatever it doesn't need.

Reviewer 9: Then why is there an opt-in at all? Is it just a temporary measure for appservices that don't know that they should apply a filter here yet?

Reviewer 2: Because spamming appservices with otherwise empty transactions isn't very great, and it still does have a performance impact. Filtering would likely make the performance issue even worse for the server-side, whereas the appservice can share the CPU cost of it.

Reviewer 9: Nit and mostly offtopic:

> Filtering would likely make the performance issue even worse for the server-side

I would expect filtering to cost a couple of orders of magnitude less than serializing and sending objects across the network. This is also assuming something about the colocation of the services which is usually close but depends on the user's setup.

hs_token: "312df522183efd404ec1cd22d2ffa4bbc76a8c1ccf541dd692eef281356bb74e"
sender_localpart: "_irc_bot"
# We want to receive EDUs
push_ephemeral: true
namespaces:
  users:
    - exclusive: true
      regex: "@_irc_bridge_.*"
  aliases:
    - exclusive: false
      regex: "#_irc_bridge_.*"
  rooms: []
```

### Changes to the /transactions/ API

The `PUT /_matrix/app/v1/transactions/{txnId}` API currently supports sending PDUs
via the `events` array.

```json
{
  "events": [
    {
      "content": {
        "membership": "join",
        "avatar_url": "mxc://domain.com/SEsfnsuifSDFSSEF#auto",
        "displayname": "Alice Margatroid"
      },
      "type": "m.room.member",
      "event_id": "$143273582443PhrSn:domain.com",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "sender": "@example:domain.com",
      "origin_server_ts": 1432735824653,
      "unsigned": {
        "age": 1234
      },
      "state_key": "@alice:domain.com"
    }
  ]
}
```

This proposal would extend the `PUT /_matrix/app/v1/transactions/` endpoint to include a new key
`ephemeral` to behave similar to the various sections of the CS API `/sync` endpoint.

```json
{

>   "ephemeral": [

Reviewer 7: Part of me is unconvinced that we should be grouping all these things together into `ephemeral`. It's different to the way `/sync` does it, which has separate `presence`, `to_device`, `device_lists` sections (though it does use `ephemeral` for typing events and read-receipts.)

That said, it is the same as the way the federation api does it.

Reviewer 3: It feels okay to treat this similar to the federation API, because we're doing a similar thing in that this API pushes the events to the AS to be consumed as events rather than a client consuming state changes (user has stopped typing, user has started typing).

Ultimately with my bridge developer hat on, this shape of API is easier to work with as it means we can consume ephemeral events similar to how we consume regular timeline events, and so far that has worked out for the ~5 bridges I've done this with.

Reviewer 4: Presence and typing/receipts can be in the same area, but it would probably make sense to separate to-device messages. Device lists and OTK counts are already separated (#3202)

Reviewer 2: Appservices are more related to federation than the sync structure, so I'd be inclined to maintain a single list.

Device lists and OTK counts are separate because of how those pieces of data are thrown around Matrix.

What is the intended benefit of splitting out to-device specifically?

Reviewer 2: For clarity: I do not want us to arbitrarily copy the federation structure wholesale as it leads to implementation complications. Appservices are still half clients.

Reviewer 4: The comments are split across 3 threads and in the third thread @richvdh said

> to-device-message types and EDU types are different namespaces

so if we don't split the to-device events, they'd have to be wrapped in an EDU like `m.direct_to_device`.

Reviewer 7: yeah sorry, I split the threads a bit here. Let's take this to https://github.com/matrix-org/matrix-doc/pull/2409#discussion_r752211109 for now.

    {

>       "type": "m.typing",

Reviewer 7: please can we use `edu_type` here, for consistency with the federation api.

Reviewer 3: Seems reasonable to me if we assume that https://github.com/matrix-org/matrix-doc/pull/2409#discussion_r752203766 is going to carry on with the federation API-like shape. :+1:

Reviewer 4: Do we *have* to copy the mistakes from the federation API too?

      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "user_ids": [
          "@alice:example.com"
        ]
      }
    },
    {
      "type": "m.receipt",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that

> existing appservices may mistake them for PDUs and could cause undefined behaviour.

Reviewer 9: ```suggestion
existing appservices may mistake them for PDUs and might behave erratically.
```

Nit: "may" can be misinterpreted as "is allowed" and we are not allowing "undefined behaviour". Also I don't think Matrix has the concept of undefined behaviour.

Reviewer: I don't think the use of "may" here is a problem, as this sentence is just saying "old implementations of application services may not expect an entry in `events` with type `m.receipt` and thus error out. Therefore let's have a new property called `ephemeral`".

"undefined behaviour" in this sentence refers to errors in older AS implementations; not in relation to the spec itself. This phrase is obviously causing confusion though, so alternative wording seems justified.

---

If this MSC weren't already implemented everywhere I may suggest that this is a weak justification for a new field, and ask for proof in existing appservice implementations of EDUs in `events` triggering errors (which implies a broken implementation...). But since we've settled on `ephemeral` in the wild - and it does make the payload a bit easier to read - I don't have an objection to keeping `ephemeral` as a separate array.

While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

`to-device` messages are a bit special as they are aimed at a particular user/device ID
combo. These events are annotated by the server with a `to_device_id` and `to_user_id`

> field at the top level of the EDU for transport to the appservice:

Reviewer 7: surely not at the top level?

Reviewer 7: actually, since the to-device messages here don't have a direct correspondence to the underlying `m.direct_to_device` events, which can contain multiple to-device messages, (see https://spec.matrix.org/v1.1/server-server-api/#send-to-device-messaging), we should probably use a separate section for to-device messages, at which point, yes can do all this at the top level.

Reviewer 7: resolving in favour of https://github.com/matrix-org/matrix-doc/pull/2409#discussion_r752211109

```json5
{

>   "type": "m.new_device",

Reviewer 7: we either need to put these in a separate section of the response, or wrap them in a `type: "m.direct_to_device"` EDU-like-thing, to avoid a namespace clash. (to-device-message types and EDU types are different namespaces)

Reviewer 7: (strong vote for a separate section from me, otherwise you've got to think of a good `type` for the pseudo-EDU. `m.direct_to_device` is misleading because it makes you think that they have the same shape as the real EDUs.)

Reviewer 3: pseudo-EDUs sound horrible, so own section feels okay to me.

Reviewer 2: Is the format at least presentable? Noting that `m.new_device` is a bad example as that is handled in [MSC3202](https://github.com/matrix-org/matrix-doc/pull/3202) specifically. A more practical example would be establishing an Olm session.

For simplicity's sake, using the same format and sticking them under a transaction top-level `to_device` array would probably be fine?

hs_token: "312df522183efd404ec1cd22d2ffa4bbc76a8c1ccf541dd692eef281356bb74e"
sender_localpart: "_irc_bot"
# We want to receive EDUs
push_ephemeral: true
namespaces:
  users:
    - exclusive: true
      regex: "@_irc_bridge_.*"
  aliases:
    - exclusive: false
      regex: "#_irc_bridge_.*"
  rooms: []
```

### Changes to the /transactions/ API

The `PUT /_matrix/app/v1/transactions/{txnId}` API currently supports sending PDUs
via the `events` array.

```json
{
  "events": [
    {
      "content": {
        "membership": "join",
        "avatar_url": "mxc://domain.com/SEsfnsuifSDFSSEF#auto",
        "displayname": "Alice Margatroid"
      },
      "type": "m.room.member",
      "event_id": "$143273582443PhrSn:domain.com",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "sender": "@example:domain.com",
      "origin_server_ts": 1432735824653,
      "unsigned": {
        "age": 1234
      },
      "state_key": "@alice:domain.com"
    }
  ]
}
```

This proposal would extend the `PUT /_matrix/app/v1/transactions/` endpoint to include a new key
`ephemeral` to behave similar to the various sections of the CS API `/sync` endpoint.

```json
{
  "ephemeral": [
    {

>       "edu_type": "m.typing",

Reviewer 4: Why do we have to copy the `type` field naming mistake from the federation API? PDUs sent to appservices are client-formatted too, so it makes no sense to use `edu_type` here just because the federation API does that.

Reviewer 2: Appservices attach themselves to a server rather than act as a client, which brings them closer to the federation level of how Matrix works. As such, it feels important (to me, imo) to acknowledge and reinforce the relationship an appservice has to its server.

Reviewer 4: I'm not sure if that's true and I don't see how this is an useful way to do that even if it is true.

* Appservices get all other events in client format. If PDUs and to-device events were sent in federation format, then it'd be understandable to use the federation format for EDUs too, but they're not, and I don't think anyone wants that either (might be fun as an optional flag like the client API has, but not by default).
* Appservices are generally based on Matrix client frameworks (matrix-{js,bot}-sdk, mautrix-{go,python}). The code for handling events seems to be partially shared with the client API in both Synapse and many appservices, so in practice, Synapse would need new code to rename the field to `edu_type`, and appservices would need new code to rename it back to `type`.

Reviewer 2: @richvdh I'd appreciate your thoughts on switching back to `type` - I'd rather not bounce between the two for years to come.

Reviewer 7: bleh. It's probably fair to say that ASes are closer to clients than servers, and as such trying to use `edu_type` for things like `m.typing` is going to require some annoying translation code on both the AS and server side.

It is annoying that it introduces an inconsistency with the federation API, and I'm not necessarily sure the "mistake" is on the federation API rather than the C-S API but :man_shrugging:.

Ultimately if those who write ASes find `type` significantly easier than `edu_type` here, then let's go with that.

Related: https://github.com/matrix-org/matrix-doc/issues/1143, https://github.com/matrix-org/matrix-doc/issues/3405.

      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that
existing appservices may mistake them for PDUs and could cause undefined behaviour.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

Note that the EDU is otherwise formatted as it would for client-server API transport.

To-device messages are a bit special as they are aimed at a particular user/device ID
combo. These events are annotated by the server with a `to_device_id` and `to_user_id`
field at the top level of the message for transport to the appservice:

```json5
{
  "type": "org.example.to_device_event_type",
  "sender": "@alice:example.com",
  "to_user_id": "@_irc_bob:example.org",
  "to_device_id": "ABCDEF123",
  "content": {
    "hello": "world"
  }
}
```

Unlike other ephemeral events, to-device messages are included at a top level `to_device`

> array in the transaction. If there are no messages to be sent, the array can be omitted.

Reviewer: Can we also omit the `ephemeral` array if there are no ephemeral events to send?

Reviewer 2: yes

Reviewer 3: 15f4582c9b00072851033e705881ecfea72c04fa

# MSC2409: Proposal to send EDUs to appservices

*Node: This proposal is a continuation of [MSC1888](https://github.com/matrix-org/matrix-doc/pull/1888)
and deprecates that one.*

The [appservice /transactions API](https://matrix.org/docs/spec/application_service/r0.1.2#put-matrix-app-v1-transactions-txnid)
currently supports pushing PDU events (regular message and state events)
however it doesn't provison for EDU events (typing, presence and more). This means that bridges cannot
react to Matrix users who send any typing or presence information in a room the service is part of.

There is an interest amongst the community to have equal bridging on both sides of a bridge, so that
read reciepts and typing notifications can be seen on the remote side. To that end, this proposal
specifies how these can be pushed to an appservice.

## Proposal

### Changes to the registration file

In order that appservices don't get flooded with EDUs, appservices have to opt-in to receive them by

> setting `push_ephemeral` to true. A registration file could look like following:

Reviewer 10: nitpick, perhaps, but since this file is describing the app service, should it be, `receive_ephemeral` rather than 'push'? (indeed, the comment below is phrased in this way too).

Reviewer 4: The field has been renamed

> setting `push_ephemeral` to true. A registration file could look like following:

Reviewer 10: I'm not 100% on this but it feels slightly confusing that `ephemeral` here matches the new `ephemeral` section the body, but it also controls the presence of the `to_device` section. Could it be clearer to just have a separate field to control whether the AS wants to_device messages (and maybe this could actually be useful for an AS to be able to subscribe to to_device message without getting presence, etc?)

Reviewer 2: In my opinion a filter would be best suited for a future MSC. Appservices already don't have an option to filter what events (PDUs) they receive, but it would be great if they could.

"Ephemeral" here is also inherited from the server-server API rather than client-server API: it includes to-device messages, but changes shape to be more useful as a representation. We could absolutely package it into `ephemeral`, but appservices will already be translating it to fit their crypto engine shape anyways - might as well give them something directly compatible.

Overall, appservices being closer to federation than clients makes things weird. We should improve them, but I don't think this is the MSC to fix everything.

      "origin_server_ts": 1432735824653,
      "unsigned": {
        "age": 1234
      },
      "state_key": "@alice:domain.com"
    }
  ]
}
```

This proposal would extend the `PUT /_matrix/app/v1/transactions/` endpoint to include a new key
`ephemeral` to behave similar to the various sections of the CS API `/sync` endpoint. The `ephemeral` key
MAY be omitted entirely if there are ephemeral no events to send.

```json
{
  "ephemeral": [
    {
      "type": "m.typing",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "user_ids": [
          "@alice:example.com"
        ]
      }
    },
    {
      "type": "m.receipt",
      "room_id": "!jEsUZKDJdhlrceRyVU:domain.com",
      "content": {
        "$1435641916114394fHBLK:matrix.org": {
          "m.read": {
            "@rikj:jki.re": {
              "ts": 1436451550453
            }
          }
        }
      }
    }
  ],
  "events": [
    // ...
  ]
}
```

The reason for a new key rather than bundling the events into `events` is that
existing appservices may mistake them for PDUs and might behave erratically.
While `events` may now be a somewhat misleading name, this is an acceptable tradeoff.

> Note that the EDU is otherwise formatted as it would for client-server API transport.

Reviewer 10: ```suggestion
Note that the EDU is otherwise formatted as it would be for client-server API transport.
```

Reviewer 4: It now explicitly says the format is C-S sync format + `room_id`

To-device messages are a bit special as they are aimed at a particular user/device ID

> combo. These events are annotated by the server with a `to_device_id` and `to_user_id`

Reviewer 10: Would maybe make more sense to wrap the to_device events in their own object which contains the metadata for this transport layer? ie.

```json
{
  "to_user_id": "@_irc_bob:example.org",
  "to_device_id": "ABCDEF123",
  {
    "type": "org.example.to_device_event_type",
    "sender": "@alice:example.com",
    "content": {
      "hello": "world"
    }
  }
}
```

...then there's no chance of overlapping keys (although obviously this would be unlikely anyway) and the receiving server can just take the inner object out and pass it on verbatim?

field at the top level of the message for transport to the appservice:

```json5
{
  "type": "org.example.to_device_event_type",
  "sender": "@alice:example.com",
  "to_user_id": "@_irc_bob:example.org",
  "to_device_id": "ABCDEF123",
  "content": {
    "hello": "world"
  }
}
```

> Unlike other ephemeral events, to-device messages are included at a top level `to_device`

Reviewer 10: Could this get a mention back where the `ephemeral` key is introduced please? ie. give a summary of what's being added and then give the detail on each.

      "sender": "@example:domain.com",
      "origin_server_ts": 1432735824653,
      "unsigned": {
        "age": 1234
      },
      "state_key": "@alice:domain.com"
    }
  ],
  "to_device": [
    {
      "type": "org.example.to_device_event_type",
      "sender": "@alice:example.com",
      "to_user_id": "@_irc_bob:example.org",
      "to_device_id": "ABCDEF123",
      "content": {
        "hello": "world"
      }
    }
  ]
}
```

### Expectations of when an EDU should be pushed to an appservice

It is not clear at face value what should be pushed to an appservice. Appservices claim
namespaces of users which registers "interest" in the rooms where those users reside, as
well as claiming namespaces of rooms for explicit interest. However, not all EDUs are
associated with a single room (presence, etc).

If the EDU is capable of being associated to a particular room, it should be sent to the
appservice under the same rules as regular events (interest in the room means sending it).
For EDUs which are not associated with a particular room, the appservice receives the EDU
if it contextually *would* apply. For example, a presence update for a user an appservice

> shares a room with (or is under the appservice's namespace) would be sent to the appservice.

Reviewer 7: Strongly recommend writing explicit rules for each type of event, rather than handwaving about EDUs in general.

(Also, note again that EDUs don't exist as a concept at all outside the federation API)

To-device messages for devices belonging to the appservice's user namespaces should always
be sent.

### Implementation detail: when to delete a to-device message

Not defined by this MSC is an explicit algorithm for when to delete a to-device message (mark
it as sent). This is left as an implementation detail, though a suggested approach is as
follows:

* If the message is sent to a user under an appservice's *exclusive* namespace, mark it as sent
  and delete it. Note that retried transactions will still need to include the message.
* If the message is sent to a user under an appservice's *inclusive* namespace, mark it as sent
  to the appservice but do not delete it until a `/sync` request is completed which includes the

>   message. Note that retried transactions will still need to include the message.

Reviewer 7: This seems to assume that a client somewhere is `/sync`ing with the same device ID as the appservice? And both devices want to receive the to-device message?

I'll note that this does not yet seem to be implemented (https://github.com/element-hq/synapse/issues/11423). Maybe that's ok since this is not a normative part of the MSC, but I'm a bit troubled that this is left undefined.

Reviewer 7: I don't really understand why we can only delete messages for users under an appservice's *exclusive* namespace. There's probably some explanation somewhere in the depths of https://github.com/matrix-org/synapse/issues/10653, but it needs explaining here imho

This approach is largely to align with how namespaces are used by appservices in practice, but
is not applicable to all scenarios (and thus is an implementation detail). The majority of known
appservices use exclusive namespaces, which typically also means that those users will not be
calling `/sync`. Because of this, the server may never get another opportunity to delete the
messages until it has confirmed that the appservice received the transaction successfully. Inclusive
namespaces are typically used when the appservice wants to impact a subset of users, but without
controlling those users explicitly. Typically, inclusive users are also calling `/sync` and so
the appservice should be CC'd on the to-device messages that would normally go down `/sync`.

## Potential issues

Determining which EDUs to transmit to the appservice could lead to quite some overhead on the
homeservers side. Additionally, more network traffic is produced, potentially straining the local
network and the appservice more. As such, appservices have to opt-in to receive EDUs.

## Security considerations

> The homeserver needs to accuratley determine which EDUs to send to the appservice, as to not leak

Reviewer 7: ```suggestion
The homeserver needs to accurately determine which EDUs to send to the appservice, as to not leak
```

any metadata about users. Particularly `m.presence` could be tricky, as no `room_id` is present in

> that EDU.

Reviewer 7: The same is true of to-device messages?
